### DISCUSSION TITLE
Am I doing it right ?

### DISCUSSION THREAD

**Discussion by GitGud31** [csells/go_router#236] [Q&A] [answered]
I Decided to use this pkg for my next big project, and I have couple of questions. I appreciate any feedback given.
Repo of the demo app: https://github.com/GitGud31/tst_gorouter_riverpod

1) Am I doing it right? 
2) How Can I use StateNotifier (riverpod) instead of ChangeNotifier in `refreshListenable`?

What I'm trying to achieve: 

  - root: '/' initial route that will either redirect to AuthScreen or HomeScreen.
  - auth: '/auth/:pid'
    - signin: '/auth/signin'
    - signup: '/auth/signup'
  {login, signup, initial} passed as params
  - home: '/home/:pid'
    - feed: '/home/feed'
    - categories: '/home/categories'
    - favorites: '/home/favorites'
    - settings: '/home/settings'
    {feed, categories, favorites, settings} passed as params
 
 I made it work, but I feed like I'm not doing thing the right way, especially in the '/' GoRoute.
 
file: root_screen.dart
 ```
final _router = GoRouter(
      debugLogDiagnostics: true,
      initialLocation: '/',
      refreshListenable: ref.watch(authStateNotifier),
      redirect: (state) {
        final _isAuthenticated =
            ref.watch(authStateNotifier.notifier).isAuthenticated;
        final _goingToAuth = (state.location == '/auth/initial' ||
            state.location == '/auth/login' ||
            state.location == '/auth/signup');

        // the user is not authenticated and not headed to /auth/:pid, they need to authenticate.
        if (!_isAuthenticated && !_goingToAuth) return '/auth/initial';

        // the user is authenticated and headed to /auth/:pid, no need to authenticate again.
        if (_isAuthenticated && _goingToAuth) return '/home/feed';

        // no need to redirect at all.
        return null;
      },
      routes: [
        GoRoute(
            name: 'root',
            path: '/',
            redirect: (state) {
              final _isAuthenticated =
                  ref.watch(authStateNotifier.notifier).isAuthenticated;

              if (_isAuthenticated) return '/home/feed';
              return '/auth/initial';
            }),
        GoRoute(
          name: 'auth',
          path: '/auth/:pid',
          builder: (_, state) {
            final pageId = state.params['pid'] ?? 'initial';
            return AuthScreen(key: state.pageKey, index: pageId);
          },
        ),
        GoRoute(
          name: 'home',
          path: '/home/:pid',
          builder: (_, state) {
            final pageId = state.params['pid'] ?? 'feed';
            return HomeScreen(key: state.pageKey, index: pageId);
          },
        ),
      ],
      errorBuilder: (_, state) => ErrorScreen(state.error!),
    );
```

**Comment by csells** [comment]
1. That looks plausible -- is it working as you expect?
2. You don't need a `ChangeNotifier` implementation, just a `Listenable`. However, `StateProvider` doesn't seem to be a `Listenable` for some reason. 

I'm curious why you don't think you did `/` correctly, @GitGud31. What are you trying to achieve?

**Reply by GitGud31** [reply]
- Yes it is working as expected.
- I tried to follow the example provided in the documentation, but instead of using using ChangeNotifier+Provider I want to use StateNotifier+StateNotifierProvider (Riverpod) but didn't know how make it work. 
-  > I'm curious why you don't think you did / correctly.

    I had to use the redirect method in the '/' **GoRoute** to listen to my auth state, when I thought the redirect method in the **GoRouter** is the one that should handle all the redirections (in this sample app case). 

-  > @GitGud31. What are you trying to achieve?

      I'd prefer if all my redirections are handled only by the redirect method in the **GoRouter**, is this possible? If so, how can I structure my routes, paths etc, to achieve this ? 
      
I appreciate any guidance as it's the first time I'm using this package (which I'm willing to start using for all my future projects :) ).

**Reply by pcelis19** [reply]
@GitGud31 did you ever figure out how to make `refreshListenable` work with a class that extends `StateNotifier`?

**Reply by GitGud31** [reply]
> @GitGud31 did you ever figure out how to make `refreshListenable` work with a class that extends `StateNotifier`?

That's unfortunately not possible, because **StateNotifier** doesn't extend or implement **Listenable** in fact it doesn't even depend of Flutter, it's pure Dart. 

You can check these links, if you want to read more: 
* [ChangeNotifier](https://api.flutter.dev/flutter/foundation/Listenable-class.html)
* [StateNotifier](https://pub.dev/documentation/state_notifier/latest/state_notifier/StateNotifier-class.html)

**Reply by danielmahon** [reply]
> @GitGud31 did you ever figure out how to make `refreshListenable` work with a class that extends `StateNotifier`?

I ended up just listening to my `AuthProvider` (StateNotifier) inside the router provider and updated and passing a local `ValueNotifier` (listenable) to GoRouter.

```dart
final routerProvider = Provider((ref) {
  final refreshListenable = ValueNotifier<AuthState?>(null);

  ref.listen<AuthState>(authProvider, (previous, next) {
    refreshListenable.value = next;
  });

  return GoRouter(
   refreshListenable: refreshListenable,
    // ...
  );
});
```

**Comment by csells** [comment, marked as answer]
You can put all of your auth redirection logic into your top-level redirect function. In fact, it's already there. Your root route can just always forward:

```dart
GoRoute(
   name: 'root',
   path: '/',
   redirect: (_) => '/home/feed',
),
```

Your top-level and route-level redirect functions will keep being called until `null` is returned. In the case that the user is authenticated and headed to `/home/feed`, then your top-level redirect will return `null` and they're head to `/home/feed` as planned. If they're not authenticated, they'll have to authenticate first before they head to `/home/feed`. I believe that's the behavior you're after, correct?

**Reply by GitGud31** [reply]
> I believe that's the behavior you're after, correct?
 
Yes pretty much

 Thanks for the exaplanations good sir!
